### ISSUE TITLE
kubernetes-dashboard and tiller won't connect to Kubernetes API on fresh install

### ISSUE THREAD

**Issue by dchambers** [rancher/rancher#7896]
**Rancher Versions:**
Server: 1.3.3
kubernetes (if applicable): 1.5.1

**Docker Version:**
1.12.3

**OS and where are the hosts located? (cloud, bare metal, etc):**
Single RancherOS host running on a VM.

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Single node rancher.

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Single Kubernetes environment.

**Steps to Reproduce:**
I've now been able to replicate this twice (both fresh installs) on two different remote VMs (one private cloud and one public cloud), but didn't see this problem when I used VirtualBox running on my own laptop.

**Results:** 
Neither `kubernetes-dashboard` nor `tiller` are able to connect to the Kubernetes API. On `kubernetes-dashboard` I see this in the logs:

```
Using HTTP port: 9090
Creating API server client for https://10.43.0.1:443
Error while initializing connection to Kubernetes apiserver. This most likely means that the cluster is misconfigured (e.g., it has invalid apiserver certificates or service accounts configuration) or the --apiserver-host param points to a server that does not exist. Reason: the server has asked for the client to provide credentials
Refer to the troubleshooting guide for more information: https://github.com/kubernetes/dashboard/blob/master/docs/user-guide/troubleshooting.md
```

and on `tiller` I see this:

```
Cannot initialize Kubernetes connection: the server has asked for the client to provide credentialspanic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x401a19]

goroutine 1 [running]:
panic(0x155bd80, 0xc42000a030)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
main.start(0x201e620, 0x2051450, 0x0, 0x0)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/cmd/tiller/tiller.go:93 +0x739
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).execute(0x201e620, 0xc42000a250, 0x0, 0x0, 0x201e620, 0xc42000a250)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:603 +0x439
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x201e620, 0x0, 0x1708aa4, 0x12)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:689 +0x367
k8s.io/helm/vendor/github.com/spf13/cobra.(*Command).Execute(0x201e620, 0x2051144, 0x16f6045)
	/home/ubuntu/.go_workspace/src/k8s.io/helm/vendor/github.com/spf13/cobra/command.go:648 +0x2b
main.main()
	/home/ubuntu/.go_workspace/src/k8s.io/helm/cmd/tiller/tiller.go:80 +0x16c
```

In both cases, this seems to be the pertinent bit:

```
Reason: the server has asked for the client to provide credentials
```

and seems to be indicating that the client is attempting to connect to the API without providing an authentication token.

Using one of the deployed containers, if I try to connect to the Kubernetes API without providing an authentication token (e.g. `curl -k https://10.43.0.1:443`) then I get an `Unauthorized` response, but I'm able to verify that I can connect to the API using the provided token as follows:

```
TOKEN_VALUE=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
curl --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt -H "Authorization: Bearer $TOKEN_VALUE" https://10.43.0.1:443
```

By running `kubectl get secrets --all-namespaces | grep service-account` I'm able to find all of the service-account tokens that exist, and by running commands like `get --namespace default -o jsonpath="{.data.token}" secret default-token-z00td | base64 -D` for each of these, I can see that the correct token has been set everywhere.

Finally, by viewing the Kubernetes config in the API for the Dashboard I can see that it has a volume definition for `/var/run/secrets/kubernetes.io/serviceaccount`, though I'm unable to confirm this within the running container as the container doesn't start.

I've also tried installing more recent versions of Rancher (`1.3.4`, `1.4.0` and `1.4.1`), and the problem also seems to occur in those, but with the downside that you no longer have the Rancher dashboard to help you debug what's going wrong in these more recent versions.

Any ideas on why this might be happening and/or where I go from here?

**Comment by dchambers** [comment]
#7241 seems to be more or less the same issue.

**Comment by domi-nik-** [comment]
I see the same issues on my install:

Rancher Versions:
Server: 1.4.1
kubernetes (if applicable): 1.5.2

Docker Version:
1.12.6

OS and where are the hosts located? (cloud, bare metal, etc):
Ubuntu 16.04 LTS virtual machine on hyper-v

Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)
HA rancher, seperate DB

Environment Type: (Cattle/Kubernetes/Swarm/Mesos)
Cattle/Kubernetes environment.

**Comment by dchambers** [comment]
I just tried installing tcpdump on the Kubernetes API container, and while I was able to observe packets coming from the kubernetes-dashboard container while it attempted to start, all of these were encrypted from the very start (TLS 1.2/1.3 I guess), so this wasn't at all helpful.

My understanding with these more recent versions of TLS is that it doesn't even help to have the keys as Diffie Hellman is now also used to provide _perfect forward secrecy_, so how does anyone debug this stuff?

**Comment by silhouette** [comment]
I had the same problems with Docker 1.12.3, for me updating docker to 1.12.6 solved it.

**Comment by dchambers** [comment]
Hi @silhouette, thanks for the suggestion, but this does not seem to have fixed things for me.

I upgraded the version of Docker used on one of the two clusters I have from `1.12.5` to `17.03.0-ce`, and it seemed like it improved things because the 'Replicas' column now showed `1`, instead of `0 of 1`, but by viewing the logs I could see it still had the same error message (`Reason: the server has asked for the client to provide credentials`) as before.

**Comment by oskapt** [comment]
@dchambers I've seen this on a fresh Rancher/k8s installation. At that time I spoke with one of my colleagues about it, and he said that the problem comes from how k8s handles authentication when not communicating over SSL.

Essentially, if the config contains auth credentials, but the target is using HTTP, kubectl/helm refuse to send the credentials. Rancher rejects the communication because it wants credentials.

I was able to get past this by putting an ELB up in environments where I could do that. In other environments I run HAProxy on the Rancher Server (for single server instances) or on a separate instance (for HA environments).  All of my communication is then to the load balancer and over SSL-backed TCP, with the load balancer talking to the Rancher servers over TCP. You want to use SSL/TCP instead of HTTP/S to avoid issues with websockets. Using self-signed certificates with k8s is a whole different issue, so if you have a real cert or can get one via LetsEncrypt or some other means, it will make your life easier.

Since making this change, all k8s environments that I spin up work perfectly the first time.

The other, less-interesting option is to remove authentication from Rancher Server. I don't recommend doing this, but if you have a test environment you can spin up, you'll see that with no auth on Rancher, k8s works over HTTP. Turn auth on in Rancher, and k8s will stop working until you communicate over an encrypted channel.

I hope this helps! If you need any asisstance with an HAProxy config or ELB guidance, I can provide additional guidance.

**Comment by dchambers** [comment]
Hi @oskapt, sorry for the delay coming back to you, and thank you so much for the offer of help here!

> Essentially, if the config contains auth credentials, but the target is using HTTP, kubectl/helm refuse to send the credentials. Rancher rejects the communication because it wants credentials.

This sounds vaguely plausible, but I don't really understand what's going on here well enough to be completely sure. One thing I am certain of is that I didn't set any auth credentials myself during the install.

> I was able to get past this by putting an ELB up in environments where I could do that. In other environments I run HAProxy on the Rancher Server (for single server instances) or on a separate instance (for HA environments). All of my communication is then to the load balancer and over SSL-backed TCP, with the load balancer talking to the Rancher servers over TCP.

I don't think this can help me in my case because Helm and the Kubernetes Dashboard are communicating directly with the Kubernetes API using the `kubernetes` service defined within the `default` namespace -- so they just use the hostname `kubernetes` to do this. Therefore, it's my understanding that there is no ELB involved. However, it does seem likely that it will be an auth issue like you say.

> The other, less-interesting option is to remove authentication from Rancher Server. I don't recommend doing this, but if you have a test environment you can spin up, you'll see that with no auth on Rancher, k8s works over HTTP.

This sounds promising! Can you please descrive how this is done and I'll give it a try?

**Comment by joshwget** [comment]
@dchambers I've tried reproducing this on a few different setups and have been unsuccessful so far. All were tested with Rancher 1.3.3.

1. AWS on RancherOS v0.7.1
2. DigitalOcean with CoreOS stable
3. Locally ( Ubuntu 16.04 )

Have you run into this problem with Rancher 1.4.x or 1.5.x?

Another note, all Docker versions above 1.12.x are currently unsupported in Kubernetes.

**Comment by dchambers** [comment]
> @dchambers I've tried reproducing this on a few different setups and have been unsuccessful so far. All were tested with Rancher 1.3.3.
>
> 1. AWS on RancherOS v0.7.1
> 2. DigitalOcean with CoreOS stable
> 3. Locally ( Ubuntu 16.04 )

Hi @joshwget, thanks so much for trying all of this.

> Have you run into this problem with Rancher 1.4.x or 1.5.x?

Yes, I've seen this problem on both 1.4.x and 1.5.x. I continue to use 1.3.3 only because it's the last version with an in-built Kubernetes dashboard.

> Another note, all Docker versions above 1.12.x are currently unsupported in Kubernetes.

Okay, let me try upgrading the cluster with Docker 1.12.3 (the other cluster had Docker 1.12.5) to 1.12.6, and see if that helps...

**Comment by joshwget** [comment]
@dchambers Can you share the Rancher registration URL used on the hosts in your setup? Are these communicating with Rancher via HTTP or HTTPS?

**Comment by dchambers** [comment]
Hi @joshwget, it's like `http://xxx.yyy.net:8080`, so **not** https.

**Comment by mdraheim** [comment]
Are those really "fresh" installs on newly provisioned hosts (VMs, droplets, whatever)?

Because I have seen this problem when trying to start from scratch on a host. Deleting the containers is not enough, I also had to delete the volumes. Bringing up a rancher k8s environment on a cleaned hosts works for me.

**Comment by neuroserve** [comment]
I stumbled upon this issue while testing Kubernetes on Rancher (v1.5.6) with https://github.com/fayazg/tritonK8ssupervisor
This sets up a new environment completely automatically with Terraform and Ansible. I have done that about five times during the last two days. The kubernetes-dashboard has not come up a single time.
The error msg is always: 

Using HTTP port: 9090
Creating API server client for https://10.43.0.1:443
Error while initializing connection to Kubernetes apiserver. This most likely means that the cluster is misconfigured (e.g., it has invalid apiserver certificates or service accounts configuration) or the --apiserver-host param points to a server that does not exist. Reason: the server has asked for the client to provide credentials
Refer to the troubleshooting guide for more information: https://github.com/kubernetes/dashboard/blob/master/docs/user-guide/troubleshooting.md

Here it tries to use https.

**Comment by dchambers** [comment]
Apologies for not sharing sooner, but here's the [work-around](https://gist.github.com/joshwget/ac657afebc6260514adbfb4de76622c1) I was provided for this issue until a proper fix is made available.

**Comment by iain17** [comment]
The only I did differently was: kubectl -n kube-system delete services --all
at the last step. Then restarted the whole k8s stack and once it was all back the dashboard was working. Thanks, @dchambers !

**Comment by deniseschannon** [comment]
We've fixed the stale secret issue as of v1.5.7 by moving how we install the secret. Fresh installs shouldn't hit this issue. The caveat is that if you upgrade to v1.5.7 and already had a broken k8s setup, it won't fix it. 

If you clean up your hosts by following the docs on what to clean up after deleting k8s, you should have a clean set up.

http://docs.rancher.com/rancher/v1.6/en/kubernetes/deleting/

**Comment by senica** [comment]
I had tried all the suggested fixes with not avail. I ended up editing the kubernetes template and turning on `Enable Kubernetes role-based access control with integrated Rancher authentication.`

And then set plane isolation to `required` and everything started working. Cost me about $75 more a month :)
